### ISSUE TITLE
Wrong type when creating a derived Int in compatible mode

### ISSUE THREAD

**Issue by ntd** [silverstripe/silverstripe-framework#6992]
The inferred type when using `Object::create` on compatible types is wrong in 3.6.0.

    class DerivedFloat extends Float {}
    class DerivedInt extends Int {}

    // Prints "DerivedInt" in SilverStripe < 3.6 and "DBInt" in 3.6.0
    echo DerivedInt::create()->class;

    // Prints "DerivedFloat" in SilverStripe < 3.6 and "DBFloat" in 3.6.0
    echo DerivedFloat::create()->class;

Using `DBInt` and `DBFloat` as base classes solves this issue.

`Int` and `Float` classes are skipped because `configFor` returns null for them so [the parent class is returned](https://github.com/silverstripe/silverstripe-framework/blob/3.6/control/injector/SilverStripeServiceConfigurationLocator.php#L25) to the injector.

Not sure what to do now... not even sure this is a bug or a planned compatibility break :)

**Comment by kinglozzer** [comment]
Ahh, 💩. This sounds like the same issue as https://github.com/silverstripe/silverstripe-framework/issues/4774, I don’t think we can solve this in 3.x.

I’m afraid you’ll have to use the following in `_config.yml`:

```yml
Injector:
  DerivedInt:
    class: DerivedInt
  DerivedFloat:
    class: DerivedFloat
```

Will that work for you @ntd?

**Comment by ntd** [comment]
Yes, I think this is a consequence of #4774.

I was wondering if it is worth a mention in the documentation. I gave the [`Object::create` docblock](https://github.com/silverstripe/silverstripe-framework/blob/3.6/core/Object.php#L115) a loook to come up with a pull request but I did not understand anything. Or my english foo sucks badly or there is something wrong there.

I personally have no problems (I'm already using `DBInt` in my project) so fill free to close if you think this is a corner case.
